helix-core/src/main/java/org/apache/helix/manager/zk/GenericBaseDataAccessorBuilder.java

```diff
@@ -82,38 +82,28 @@ protected RealmAwareZkClient createZkClient(RealmAwareZkClient.RealmMode realmMo
     switch (realmMode) {
       case MULTI_REALM:
         try {
-          if (_zkClientType == ZkClientType.DEDICATED) {
-            // Use a realm-aware dedicated zk client
-            zkClient = DedicatedZkClientFactory.getInstance()
-                .buildZkClient(connectionConfig, clientConfig);
-          } else if (_zkClientType == ZkClientType.SHARED) {
-            // Use a realm-aware shared zk client
-            zkClient =
-                SharedZkClientFactory.getInstance().buildZkClient(connectionConfig, clientConfig);
-          } else {
-            zkClient = new FederatedZkClient(connectionConfig, clientConfig);
-          }
-        } catch (IOException | InvalidRoutingDataException | IllegalStateException e) {
+          zkClient = new FederatedZkClient(connectionConfig, clientConfig);
```

Comment:
If you think this is a bug we should fix, then we need to make clear definition of what is single realm and what is multi realm. My original thought was if you are using sharding key to determine the ZK address then it is mutli realm.

Then for dedicated zkclient, and shared zkclient, they can still use sharding key for searching zk address. Then it will be meaningless for dedicated/shared Zkclient to have the sharding key and they will only accept the ZK address. This conflict with what we have changed.

Reply:
> If you think this is a bug we should fix, then we need to make clear definition of what is single realm and what is multi realm. My original thought was if you are using sharding key to determine the ZK address then it is mutli realm.

"Multi-realm" means it needs to be able to talk to different realms. "Single-realm" means it only connects to one realm.
If you're using a sharding key to locate which ZK to talk to, then it's "realm-aware", not multi-realm.

> Then for dedicated zkclient, and shared zkclient, they can still use sharding key for searching zk address. Then it will be meaningless for dedicated/shared Zkclient to have the sharding key and they will only accept the ZK address. This conflict with what we have changed.

RealmAware DedicatedZkClient and SharedZkClient do not require the ZK address. They instead require the ZK path sharding key and only connect to the ZK realm that the given ZK path sharding key maps to. Again, they fall under the single-realm use case.

@dasahcc Does this clarify?

Reply:
Sounds good. We need to clear deliver this to users. Then for the change. I am totally fine with it.